modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java

```diff
@@ -0,0 +1,258 @@
+                if (updateTaskState(name)) {
+                    logger.info("updated geoip database [" + name + "]");
+                }
+            }
+        } catch (Exception e) {
+            logger.error("error updating geoip database [" + name + "]", e);
+        }
+    }
+
+    //visible for testing
+    protected void updateTimestamp(Map<String, Metadata> currentDatabases, String name) {
+        Metadata old = currentDatabases.get(name);
+        currentDatabases.put(name, new Metadata(System.currentTimeMillis(), old.getFirstChunk(), old.getLastChunk(), old.getMd5()));
+        logger.info("geoip database [" + name + "] is up to date, updated timestamp");
+        updateTaskState(name);
+    }
+
+    boolean updateTaskState(String name) {
+        PlainActionFuture<PersistentTask<?>> future = PlainActionFuture.newFuture();
+        persistentTask.updatePersistentTaskState(state, future);
+        try {
+            state = ((GeoIpTaskState) future.actionGet().getState()).copy();
+            return true;
+        } catch (Exception e) {
+            logger.error("can not update state during geoip database [" + name + "] update", e);
+            return false;
+        }
+    }
+
+    //visible for testing
+    int indexChunks(String name, InputStream is, int chunk) throws IOException {
+        while (true) {
+            byte[] buf = getChunk(is);
+            if (buf.length == 0) {
+                break;
+            }
+            client.prepareIndex(DATABASES_INDEX).setId(name + "_" + chunk)
+                .setSource(XContentType.SMILE, "name", name, "chunk", chunk, "data", buf)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .setWaitForActiveShards(ActiveShardCount.ALL)
+                .get();
+            chunk++;
+        }
+        return chunk;
+    }
+
+    //visible for testing
+    byte[] getChunk(InputStream is) throws IOException {
+        byte[] buf = new byte[MAX_CHUNK_SIZE];
+        int chunkSize = 0;
+        while (chunkSize < MAX_CHUNK_SIZE) {
+            int read = is.read(buf, chunkSize, MAX_CHUNK_SIZE - chunkSize);
+            if (read == -1) {
+                break;
+            }
+            chunkSize += read;
+        }
+        if (chunkSize < MAX_CHUNK_SIZE) {
+            buf = Arrays.copyOf(buf, chunkSize);
+        }
+        return buf;
+    }
+
+    //for testing only
+    void setPersistentTask(AllocatedPersistentTask persistentTask) {
+        this.persistentTask = persistentTask;
+    }
+
+    //for testing only
+    void setState(GeoIpTaskState state) {
+        this.state = state;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask allocatedTask, PersistentTaskParams params, PersistentTaskState state) {
+        if (scheduled != null) {
+            scheduled.cancel();
+        }
+        if (enabled) {
```

Comment:
I think we should always check here whether the node task has stopped for some reason, so that we don't have multiple instances running of the same persistent task (since we rerun from a thread pool without checking the node task's state).

Reply:
Something like `if (enabled || allocatedTask.isCancelled() || allocatedTask.isCompleted()) {`

The issue is the `isCompleted()` has protected access, I think we should introduce a custom `AllocatedPersistentTask` instance and wrap the downloader logic into that, which has other benefits
like having a reference to the actual `PersistentTasksService`, being able to execute `scheduled.cancel()`
when the downloader get cancelled / stopped. Also we can then overwrite `getStatus()` method on the Task class, for reporting stats via the task api (low level stats that we wouldn't have to persist in cluster state, that might be useful for debugging purposes).

Reply:
I've moved logic to `AllocatedPersistentTask` and it does make the code simpler and safer I think. Thanks for suggestion!

Reply:
👍 Looks good!